File: ios/WhoWouldWin/Views/Tournament/TournamentBracketDiagram.swift
import SwiftUI

/// A compact visualization of a tournament bracket.
/// Each round is a column; each matchup is a small card showing the two fighters
/// and, if resolved, the winner highlighted. Used in BracketPreviewView,
/// RoundWagerView headers, and TournamentCompleteView.
struct TournamentBracketDiagram: View {
    let bracket: Bracket
    /// Optional round index to highlight (e.g. the current round)
    var highlightedRoundIndex: Int? = nil
    /// When true, the diagram is scrollable horizontally.
    var scrollable: Bool = true

    @Environment(\.horizontalSizeClass) private var sizeClass
    private var isIPad: Bool { sizeClass == .regular }

    var body: some View {
        let content = HStack(alignment: .center, spacing: isIPad ? 18 : 12) {
            ForEach(Array(bracket.rounds.enumerated()), id: \.offset) { (roundIdx, round) in
                VStack(spacing: isIPad ? 14 : 10) {
                    Text(roundLabel(roundIdx))
                        .font(Kids.fredoka(isIPad ? 13 : 10, weight: .bold))
                        .foregroundColor(Kids.ink)
                        .tracking(1)
                        .padding(.horizontal, isIPad ? 12 : 8).padding(.vertical, isIPad ? 5 : 3)
                        .background(
                            Capsule().fill(Kids.sun)
                                .overlay(Capsule().stroke(Kids.ink, lineWidth: 1.5))
                        )

                    if round.isEmpty {
                        ForEach(0..<max(1, placeholderCount(for: roundIdx)), id: \.self) { _ in
                            placeholderCard
                        }
                    } else {
                        ForEach(round) { matchup in
                            matchupCard(matchup)
                        }
                    }
                }
                .opacity(highlightedRoundIndex == nil || highlightedRoundIndex == roundIdx ? 1.0 : 0.55)
            }
        }
        .padding(.horizontal, isIPad ? 14 : 8)
        .padding(.vertical, isIPad ? 10 : 6)

        if scrollable {
            ScrollView(.horizontal, showsIndicators: false) { content }
        } else {
            content
        }
    }

    // MARK: - Helpers

    private func roundLabel(_ roundIdx: Int) -> String {
        let size: BracketSize
        switch bracket.rounds.count {
        case 2: size = .four
        case 3: size = .eight
        case 4: size = .sixteen
        default: size = .four
        }
        return size.roundName(for: roundIdx).uppercased()
    }

    private func placeholderCount(for roundIdx: Int) -> Int {
        let total = bracket.rounds.count
        let remaining = total - roundIdx
        switch remaining {
        case 1: return 1
        case 2: return 2
        case 3: return 4
        case 4: return 8
        default: return 1
        }
    }

    @ViewBuilder
    private func matchupCard(_ matchup: Matchup) -> some View {
        VStack(spacing: isIPad ? 5 : 3) {
            fighterRow(matchup.fighter1,
                       isWinner: matchup.winningFighter?.id == matchup.fighter1.id,
                       isLoser:  matchup.losingFighter?.id == matchup.fighter1.id)
            Rectangle()
                .fill(Kids.ink.opacity(0.15))
                .frame(height: 1)
            fighterRow(matchup.fighter2,
                       isWinner: matchup.winningFighter?.id == matchup.fighter2.id,
                       isLoser:  matchup.losingFighter?.id == matchup.fighter2.id)
        }
        .padding(.vertical, isIPad ? 10 : 6)
        .padding(.horizontal, isIPad ? 12 : 8)
        .frame(width: isIPad ? 188 : 128)
        .background(
            RoundedRectangle(cornerRadius: isIPad ? 16 : 12, style: .continuous)
                .fill(.white)
                .overlay(RoundedRectangle(cornerRadius: isIPad ? 16 : 12, style: .continuous).stroke(Kids.ink, lineWidth: 2))
        )
        .shadow(color: Kids.ink.opacity(0.06), radius: 0, x: 0, y: 2)
    }

    @ViewBuilder
    private func fighterRow(_ animal: Animal, isWinner: Bool, isLoser: Bool) -> some View {
        HStack(spacing: isIPad ? 7 : 5) {
            DiagramMini(animal: animal, isIPad: isIPad)
            Text(animal.name)
                .font(Kids.fredoka(isIPad ? 13 : 10, weight: .bold))
                .foregroundColor(isLoser ? Kids.inkSoft.opacity(0.5) : Kids.ink)
                .lineLimit(1)
                .minimumScaleFactor(0.65)
                .strikethrough(isLoser)
            Spacer(minLength: 0)
            if isWinner {
                Text("👑")
                    .font(.system(size: isIPad ? 15 : 11))
            }
        }
    }

    private var placeholderCard: some View {
        VStack(spacing: isIPad ? 5 : 3) {
            HStack {
                Text("???").font(Kids.fredoka(isIPad ? 13 : 10, weight: .bold)).foregroundColor(Kids.inkSoft.opacity(0.5))
                Spacer()
            }
            Rectangle().fill(Kids.ink.opacity(0.1)).frame(height: 1)
            HStack {
                Text("???").font(Kids.fredoka(isIPad ? 13 : 10, weight: .bold)).foregroundColor(Kids.inkSoft.opacity(0.5))
                Spacer()
            }
        }
        .padding(.vertical, isIPad ? 10 : 6)
        .padding(.horizontal, isIPad ? 12 : 8)
        .frame(width: isIPad ? 188 : 128)
        .background(
            RoundedRectangle(cornerRadius: 12, style: .continuous)
                .fill(Color(hex: "#F7F2FF"))
        )
        .overlay(
            RoundedRectangle(cornerRadius: 12, style: .continuous)
                .stroke(Kids.ink.opacity(0.15), style: StrokeStyle(lineWidth: 1.5, dash: [3, 3]))
        )
    }
}

private struct DiagramMini: View {
    let animal: Animal
    let isIPad: Bool
    private var bundledImage: UIImage? {
        guard let name = animal.creatureAssetName else { return nil }
        return UIImage(named: name)
    }
    var body: some View {
        let outer: CGFloat = isIPad ? 30 : 22
        let inner: CGFloat = isIPad ? 22 : 16
        ZStack {
            Circle().fill(.white)
                .overlay(Circle().stroke(Kids.ink, lineWidth: 1.5))
                .frame(width: outer, height: outer)
            Group {
                if let ui = bundledImage {
                    Image(uiImage: ui).resizable().scaledToFill()
                } else {
                    Text(animal.emoji).font(.system(size: isIPad ? 16 : 12))
                }
            }
            .frame(width: inner, height: inner)
            .clipShape(Circle())
        }
    }
}

File: android/app/src/main/java/com/whowouldin/whowouldwin/ui/screens/tournament/TournamentBracketDiagram.kt
package com.whowouldin.whowouldwin.ui.screens.tournament

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.whowouldin.whowouldwin.model.Animal
import com.whowouldin.whowouldwin.model.Bracket
import com.whowouldin.whowouldwin.model.BracketSize
import com.whowouldin.whowouldwin.model.Matchup
import com.whowouldin.whowouldwin.ui.components.AnimalAvatar
import com.whowouldin.whowouldwin.ui.theme.BrandTheme
import com.whowouldin.whowouldwin.ui.theme.bungee

/**
 * Port of iOS TournamentBracketDiagram.
 *
 * Horizontal list of round columns, each with compact matchup cards.
 * Shows winners highlighted with a crown + loser name strikethrough.
 */
@Composable
fun TournamentBracketDiagram(
    bracket: Bracket,
    modifier: Modifier = Modifier,
    highlightedRoundIndex: Int? = null,
    scrollable: Boolean = true,
) {
    val content: @Composable () -> Unit = {
        Row(
            modifier = Modifier.padding(horizontal = 8.dp, vertical = 6.dp),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            bracket.rounds.forEachIndexed { roundIdx, round ->
                val alpha = if (highlightedRoundIndex == null || highlightedRoundIndex == roundIdx) 1f else 0.55f
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(10.dp),
                    modifier = Modifier
                        .padding(bottom = 2.dp),
                ) {
                    Text(
                        text = roundLabel(bracket, roundIdx),
                        style = bungee(11).copy(
                            color = Color.White.copy(alpha = 0.75f * alpha),
                            letterSpacing = 1.sp,
                        ),
                        modifier = Modifier.padding(bottom = 2.dp),
                    )
                    if (round.isEmpty()) {
                        repeat(placeholderCount(bracket, roundIdx).coerceAtLeast(1)) {
                            PlaceholderCard(alpha)
                        }
                    } else {
                        round.forEach { m -> MatchupCard(m, alpha) }
                    }
                }
            }
        }
    }

    if (scrollable) {
        Box(modifier.horizontalScroll(rememberScrollState())) { content() }
    } else {
        Box(modifier) { content() }
    }
}

private fun roundLabel(b: Bracket, idx: Int): String {
    val size = when (b.rounds.size) {
        2 -> BracketSize.FOUR
        3 -> BracketSize.EIGHT
        4 -> BracketSize.SIXTEEN
        else -> BracketSize.FOUR
    }
    return size.roundName(idx).uppercase()
}

private fun placeholderCount(b: Bracket, idx: Int): Int {
    val remaining = b.rounds.size - idx
    return when (remaining) {
        1 -> 1; 2 -> 2; 3 -> 4; 4 -> 8; else -> 1
    }
}

@Composable
private fun MatchupCard(m: Matchup, alpha: Float) {
    Column(
        modifier = Modifier
            .width(120.dp)
            .background(Color.White.copy(alpha = 0.08f * alpha), RoundedCornerShape(10.dp))
            .border(1.dp, Color.White.copy(alpha = 0.25f * alpha), RoundedCornerShape(10.dp))
            .padding(vertical = 6.dp, horizontal = 8.dp),
        verticalArrangement = Arrangement.spacedBy(2.dp),
    ) {
        FighterRow(
            animal = m.fighter1,
            isWinner = m.winningFighter?.id == m.fighter1.id,
            isLoser = m.losingFighter?.id == m.fighter1.id,
            alpha = alpha,
        )
        Box(Modifier.fillMaxWidth().height(1.dp).background(Color.White.copy(alpha = 0.25f * alpha)))
        FighterRow(
            animal = m.fighter2,
            isWinner = m.winningFighter?.id == m.fighter2.id,
            isLoser = m.losingFighter?.id == m.fighter2.id,
            alpha = alpha,
        )
    }
}

@Composable
private fun FighterRow(animal: Animal, isWinner: Boolean, isLoser: Boolean, alpha: Float) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(5.dp),
        modifier = Modifier.fillMaxWidth(),
    ) {
        AnimalAvatar(animal = animal, size = 20.dp, cornerRadius = 5.dp)
        Text(
            text = animal.name,
            style = bungee(11).copy(
                color = (if (isLoser) Color.White.copy(alpha = 0.35f) else Color.White).copy(alpha = alpha),
                textDecoration = if (isLoser) TextDecoration.LineThrough else TextDecoration.None,
            ),
            maxLines = 1,
            modifier = Modifier.weight(1f),
        )
        if (isWinner) {
            Icon(
                imageVector = Icons.Filled.Star,
                contentDescription = null,
                tint = BrandTheme.gold.copy(alpha = alpha),
                modifier = Modifier.size(10.dp),
            )
        }
    }
}

@Composable
private fun PlaceholderCard(alpha: Float) {
    Column(
        modifier = Modifier
            .width(120.dp)
            .background(Color.White.copy(alpha = 0.04f * alpha), RoundedCornerShape(10.dp))
            .border(1.dp, Color.White.copy(alpha = 0.12f * alpha), RoundedCornerShape(10.dp))
            .padding(vertical = 6.dp, horizontal = 8.dp),
        verticalArrangement = Arrangement.spacedBy(2.dp),
    ) {
        Text("? ? ?", style = bungee(11).copy(color = Color.White.copy(alpha = 0.35f * alpha)))
        Spacer(Modifier.height(2.dp))
        Text("? ? ?", style = bungee(11).copy(color = Color.White.copy(alpha = 0.35f * alpha)))
    }
}
